data/tutorials/language/0lg_08_env.md

```diff
@@ -0,0 +1,557 @@
+
+Here is the translation from the first kind to the second kind:
+```ocaml
+# fun f -> fun (x, y) -> f x y;;
+- : ('a -> 'b -> 'c) -> 'a * 'b -> 'c = <fun>
+```
+
+Here is the reverse translation:
+```ocaml
+# fun f -> fun x y -> f (x, y);;
+- : ('a * 'b -> 'c) -> 'a -> 'b -> 'c = <fun>
+```
+
+These translations are attributed to the 20th-century logician [Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry). The second translation is called currying and the first is called uncurrying.
+
+From a typing perspective, this means that for any types `a`, `b`, and `c`, the following types are equivalent:
+- `a -> (b -> c)` &mdash; curried function type
+- `a * b -> c` &mdash; uncurried function type
+
+We will not dive any deeper into the details. This equivalence can be formally defined using _ad-hoc_ mathematics.
+
+It is very rare to need to use the functions above. However, it is important to understand that changing a function may not need any code refactoring. A function can be given several equivalent forms and changed into another, either using refactoring or using a higher-order function. Since functions are values, currying and uncurrying are operations on those values.
+
+In practice, curried functions are the default form function should take.
+- Allows partial application
+- Less editing, no parenthesis and comas
+- No pattern matching over a tuple
+
+### Functions With Side-Effects
+
+With respect to its type, a function is expected to process input data from its domain and produce a result data from its codomain.
+
+However, some functions either take input data outside of their declared domain or produce data outside of their codomain. These out-of-signature data are called effects, or side effects. Input and output are the most common forms of effects. Input is out-of-domain data and output out-of-codomain data. However, the result of functions returning random numbers or the current time is influenced by external factors, which is also called an effect. The external factor is out-of-domain input. Similarly, any observable phenomena triggered by the computation of a function is out-of-codomain output.
+
+In practice, what is considered an effect is an engineering choice. In most circumstances, I/O are considered as effects, unless they are ignored. Electromagnetic radiation emitted by the processor when computing a function isn't usually considered a relevant effect, except in some security-sensitive contexts. In the OCaml community, as well as in the wider functional programming community, functions are often said to be either pure or impure. The former does not have side effects, the latter does. It is important to remember this always assumes some sort of context, any computation has effects, and what is considered relevant is a choice.
```

Comment:
I'd like to work on this sentence together for clarification/syntax: "It is important to remember this always assumes some sort of context, any computation has effects, and what is considered relevant is a choice."